SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,278 @@
+	}
+}
+
+void ULayeredLBStrategy::Init()
+{
+	Super::Init();
+
+	VirtualWorkerId CurrentVirtualWorkerId = SpatialConstants::INVALID_VIRTUAL_WORKER_ID + 1;
+
+	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	{
+		check(Settings->bEnableUnrealLoadBalancer);
+		const ASpatialWorldSettings* WorldSettings = GetWorld() ? Cast<ASpatialWorldSettings>(GetWorld()->GetWorldSettings()) : nullptr;
+
+		if (WorldSettings == nullptr)
+		{
+			UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, WorldSettings should inherit from SpatialWorldSettings to get the load balancing strategy."));
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+			AddStrategyForLayer(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+			return;
+		}
+
+		// This will be uncommented after the next PR.
+		// For each Layer, add a LB Strategy for that layer.
+// 			TMap<FName, FLBLayerInfo> WorkerLBLayers = WorldSettings->WorkerLBLayers;
+// 
+// 			for (const TPair<FName, FLayerInfo>& Layer : Settings->WorkerLayers)
+// 			{
+// 				FName LayerName = Layer.Key;
+// 
+// 				// Look through the WorldSettings to find the LBStrategy type for this layer.
+// 				if (!WorkerLBLayers.Contains(LayerName))
+// 				{
+// 					UE_LOG(LogLayeredLBStrategy, Error, TEXT("Layer %s does not have a defined LBStrategy in the WorldSettings. It will not be simulated."), *(LayerName.ToString()));
+// 					continue;
+// 				}
+// 
+// 				UAbstractLBStrategy* LBStrategy = NewObject<UAbstractLBStrategy>(this, WorkerLBLayers[LayerName].LoadBalanceStrategy);
+// 				AddStrategyForLayer(LayerName, LBStrategy);
+// 
+// 				for (const TSoftClassPtr<AActor>& ClassPtr : Layer.Value.ActorClasses)
+// 				{
+// 					ClassPathToLayer.Add(ClassPtr, LayerName);
+// 				}
+// 			}
+
+		// Finally, add the default layer.
+		if (WorldSettings->LoadBalanceStrategy == nullptr)
+		{
+			UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, there must be a LoadBalancing strategy set. Using a 1x1 grid."));
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+			AddStrategyForLayer(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+		}
+		else
+		{
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UAbstractLBStrategy>(this, WorldSettings->LoadBalanceStrategy);
+			AddStrategyForLayer(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+		}
+	}
+}
+
+void ULayeredLBStrategy::SetLocalVirtualWorkerId(VirtualWorkerId InLocalVirtualWorkerId)
+{
+	LocalVirtualWorkerId = InLocalVirtualWorkerId;
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		Elem.Value->SetLocalVirtualWorkerId(InLocalVirtualWorkerId);
+	}
+}
+
+bool ULayeredLBStrategy::ShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to relinquish authority for Actor %s."), *AActor::GetDebugName(&Actor));
```

Comment:
Can we upgrade to error? I think seeing this means some part of world initialisation won't be able to make the decision it needs and some Actor could  end up as authoritative when it should be simulated proxy or vice versa?

Reply:
As discussed with Matt, I'd rather wait for later and make one big pass over all of the initialization order.